### PR TITLE
Reduced unnecessary block updates and some code cleanup

### LINKED ISSUE #3145: The generator creates grass in undesired areas
### Expected behavior
The generator is expected to create grass on a block only when air is above.

### Actual behavior
The generator creates grass under tree trunks and snow. that grass decays into dirt, sending a non-negligible amount needless packets for several minutes upon entering a newly generated area.

#### Benchmark
I created a fresh world and logged in. about 32 block updates were being sent per second! This number gradually decreased until reaching zero. The log is attached here:
[out.txt](https://github.com/cuberite/cuberite/files/222337/out.txt)

Related: #3129 , #3142 , #3110 , #3021

### PR DESCRIPTION
**Pull request by LogicParrot**
- Chunk generator no longer generates grass under tree trunks. This reduces dirty chunks. Closes #3145
- Water simulator no longer marks chunks dirty for no good reason.
- Leaves are now created with the proper metas, no hundreds of updates are sent to the client (#3129 had a hotfix for this, but here it is done properly)
- Closes a portion of https://github.com/cuberite/cuberite/issues/3110

### CODE REVIEW

**Comment by Seadragon91** [comment]
Would this also fix #2766?

**Comment by LogicParrot** [comment]
Very possibly. And it may fix this too: #3021 well, at least for stationary players. (I don't know well/bad sending full chunks is).

File: src/Chunk.cpp [hunk 1]
```diff
@@ -1536,8 +1535,17 @@ void cChunk::FastSetBlock(int a_RelX, int a_RelY, int a_RelZ, BLOCKTYPE a_BlockT
 		return;
 	}
 
-	MarkDirty();
-	m_IsRedstoneDirty = true;
+	bool ReplacingLiquids = ( \
```
Comment (tigerw) [inline src/Chunk.cpp]: > \

o.0

File: src/Entities/Entity.cpp [hunk 2]
```diff
@@ -228,6 +228,7 @@ void cEntity::Destroy(bool a_ShouldBroadcast)
 			this->GetUniqueID(), this->GetClass(),
 			ParentChunk->GetPosX(), ParentChunk->GetPosZ()
 		);
+		LOGD("%g %g %g", GetPosX(), GetPosY(), GetPosZ());
```
Comment (tigerw) [inline src/Entities/Entity.cpp]: And the comment to self in...

File: src/Chunk.cpp (hunk 1 above)
Comment (109C) [inline src/Chunk.cpp]: >(\
/)

File: src/Entities/Entity.cpp (hunk 2 above)
Comment (LogicParrot) [inline src/Entities/Entity.cpp]: Lol !!

File: src/Generating/Trees.cpp [hunk 3]
```diff
@@ -341,24 +341,24 @@ void GetSmallAppleTreeImage(int a_BlockX, int a_BlockY, int a_BlockZ, cNoise & a
 	// 2 BigO2 + corners layers:
 	for (int i = 0; i < 2; i++)
 	{
-		PushCoordBlocks (a_BlockX, Hei, a_BlockZ, a_OtherBlocks, BigO2, ARRAYCOUNT(BigO2), E_BLOCK_LEAVES, E_META_LEAVES_APPLE);
-		PushCornerBlocks(a_BlockX, Hei, a_BlockZ, a_Seq, a_Noise, 0x5000000 - i * 0x10000000, a_OtherBlocks, 2, E_BLOCK_LEAVES, E_META_LEAVES_APPLE);
+		PushCoordBlocks (a_BlockX, Hei, a_BlockZ, a_OtherBlocks, BigO2, ARRAYCOUNT(BigO2), E_BLOCK_LEAVES, E_META_LEAVES_APPLE_CHECK_DECAY);
```
Comment (tigerw) [inline src/Generating/Trees.cpp]: Hmmm. For this and for all below, the MC wiki states that "if the check decay bit is set, the leaf is checked for decay. The bit is set when adjacent blocks are changed." Am I misunderstanding, or is our leaf decay code broken, or was the generator fine in the first place?

File: src/Entities/Entity.cpp (hunk 2 above)
Comment (tigerw) [inline src/Entities/Entity.cpp]: ...now.

File: src/Generating/Trees.cpp (hunk 3 above)
Comment (LogicParrot) [inline src/Generating/Trees.cpp]: We might be having some flipped roles... Let me check this.

Comment (LogicParrot) [inline src/Generating/Trees.cpp]: We had it flipped. I fixed it. I propose that we should not inform the clients of these meta updates, at least until all forests "normalize" for most people, and possibly forever.

File: src/Generating/StructGen.cpp [hunk 4]
```diff
@@ -183,6 +183,14 @@ void cStructGenTrees::ApplyTreeImage(
 				CASE_TREE_OVERWRITTEN_BLOCKS:
 				{
 					a_ChunkDesc.SetBlockTypeMeta(itr->m_RelX, itr->m_RelY, itr->m_RelZ, itr->m_BlockType, itr->m_BlockMeta);
+					// If grass is below our tree, turn it to dirt
```
Comment (tigerw) [inline src/Generating/StructGen.cpp]: Tree! Tree? My dear LogicParrot, this is a *StructGen*.

Still, I dislike the whole commit-fetch-rebase-push, especially for little things like comments, so I shan't mind if you don't bother.

File: src/Generating/StructGen.cpp [hunk 5]
```diff
@@ -183,6 +183,14 @@ void cStructGenTrees::ApplyTreeImage(
...
+					if (
+						(cBlockInfo::IsSolid(itr->m_BlockType)) &&
```
Comment (tigerw) [inline src/Generating/StructGen.cpp]: Could it be a good idea to pull the grass-to-dirt check out of BlockDirt and use that? I remember something about there being water checks too?

File: src/Generating/StructGen.cpp [hunk 6]
```diff
@@ -183,6 +183,14 @@ void cStructGenTrees::ApplyTreeImage(
 				CASE_TREE_OVERWRITTEN_BLOCKS:
```
Comment (tigerw) [inline src/Generating/StructGen.cpp]: I'd just like to take a moment to appreciate the peculiarity of this statement. It is truly a *case* of unnecessary macro use.

File: src/Generating/Trees.cpp (hunk 3 above)
Comment (tigerw) [inline src/Generating/Trees.cpp]: I remain sceptical about meta updates. The client seems to send a position update once every tick from previous observation. If I am not wrong in this regard, then the connection can handle it fine. I would be interested to see a breakdown of all sent packet types, somehow. But, what do I know :P

File: src/Generating/StructGen.cpp (hunk 4 above)
Comment (LogicParrot) [inline src/Generating/StructGen.cpp]: What's wrong? This is a tree.

File: src/Generating/StructGen.cpp (hunk 5 above)
Comment (LogicParrot) [inline src/Generating/StructGen.cpp]: Water can decay grass. But since this is a tree being generated on grass, there is no water here.

**Comment by LogicParrot** [comment]
Can we merge this?
@tigerw I don't quite understand what you meant by the tree comment. Would you mind explaining? Is there anything wrong?

**Comment by tigerw** [comment]
Of couse you can merge it. Never mind about the comment about tree comments, it was just a thought.

**Pull request merged**